File: app/src/main/java/com/bccommittee/adapter/FundAdapter.java
package com.bccommittee.adapter;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.bccommittee.R;
import com.bccommittee.model.Fund;
import com.bccommittee.util.CurrencyUtil;

import java.util.ArrayList;
import java.util.List;

public class FundAdapter extends RecyclerView.Adapter<FundAdapter.FundViewHolder> {

    public interface OnFundClickListener {
        void onFundClick(Fund fund);
    }

    private final Context context;
    private final OnFundClickListener listener;
    private List<Fund> funds = new ArrayList<>();

    public FundAdapter(Context context, OnFundClickListener listener) {
        this.context = context;
        this.listener = listener;
    }

    public void setFunds(List<Fund> funds) {
        this.funds = funds;
        notifyDataSetChanged();
    }

    @NonNull
    @Override
    public FundViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(R.layout.item_fund, parent, false);
        return new FundViewHolder(view);
    }

    @Override
    public void onBindViewHolder(@NonNull FundViewHolder holder, int position) {
        Fund fund = funds.get(position);
        holder.bind(fund);
        holder.itemView.setOnClickListener(v -> listener.onFundClick(fund));
    }

    @Override
    public int getItemCount() {
        return funds.size();
    }

    static class FundViewHolder extends RecyclerView.ViewHolder {
        TextView tvFundName, tvFundMeta, tvStatus, tvMonthly,
                 tvInterestRate, tvOverdueCount, tvClosedDate;
        View viewDot;
        FrameLayout frameDot;

        FundViewHolder(View v) {
            super(v);
            tvFundName     = v.findViewById(R.id.tv_fund_name);
            tvFundMeta     = v.findViewById(R.id.tv_fund_meta);
            tvStatus       = v.findViewById(R.id.tv_status);
            tvMonthly      = v.findViewById(R.id.tv_monthly);
            tvInterestRate = v.findViewById(R.id.tv_interest_rate);
            tvOverdueCount = v.findViewById(R.id.tv_overdue_count);
            tvClosedDate   = v.findViewById(R.id.tv_closed_date);
            viewDot        = v.findViewById(R.id.view_dot);
            frameDot       = v.findViewById(R.id.frame_dot);
        }

        void bind(Fund fund) {
            tvFundName.setText(fund.getName());
            tvFundMeta.setText(fund.getTotalMembers() + " members · "
                    + CurrencyUtil.format(fund.getMonthlyAmount()) + "/mo");
            tvMonthly.setText(CurrencyUtil.format(fund.getMonthlyAmount()));
            tvInterestRate.setText(fund.getInterestRate() + "% / mo");
            tvOverdueCount.setText(String.valueOf(fund.getOverdueCount()));

            // Dot color
            try {
                int dotColor = Color.parseColor(
                        fund.getDotColor() != null ? fund.getDotColor() : "#1F5C3A");
                GradientDrawable dotBg = new GradientDrawable();
                dotBg.setShape(GradientDrawable.OVAL);
                dotBg.setColor(dotColor);
                viewDot.setBackground(dotBg);

                GradientDrawable frameBg = new GradientDrawable();
                frameBg.setShape(GradientDrawable.RECTANGLE);
                frameBg.setCornerRadius(20f);
                frameBg.setColor(Color.argb(34, Color.red(dotColor),
                        Color.green(dotColor), Color.blue(dotColor)));
                frameBg.setStroke(4, Color.argb(85, Color.red(dotColor),
                        Color.green(dotColor), Color.blue(dotColor)));
                frameDot.setBackground(frameBg);
            } catch (Exception ignored) {}

            // Status pill
            if (fund.isClosed()) {
                tvStatus.setText("Closed");
                tvStatus.setBackgroundResource(R.drawable.pill_gray_bg);
                tvStatus.setTextColor(itemView.getContext().getColor(R.color.closed_text));
                tvClosedDate.setVisibility(View.VISIBLE);
                tvClosedDate.setText("Closed: " + (fund.getClosedDate() != null ? fund.getClosedDate() : ""));
                tvMonthly.setText("—");
                tvInterestRate.setText("—");
            } else if (fund.isOverdue()) {
                tvStatus.setText("⚠ Overdue");
                tvStatus.setBackgroundResource(R.drawable.pill_orange_bg);
                tvStatus.setTextColor(itemView.getContext().getColor(R.color.color_orange));
                tvClosedDate.setVisibility(View.GONE);
            } else {
                tvStatus.setText("✓ All Paid");
                tvStatus.setBackgroundResource(R.drawable.pill_green_bg);
                tvStatus.setTextColor(itemView.getContext().getColor(R.color.color_green));
                tvClosedDate.setVisibility(View.GONE);
            }
        }
    }
}

File: app/src/main/kotlin/com/khaga/ksociety/adapter/FundAdapter.kt
package com.khaga.ksociety.adapter

import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.khaga.ksociety.R
import com.khaga.ksociety.databinding.ItemFundBinding
import com.khaga.ksociety.model.Fund
import com.khaga.ksociety.util.CurrencyUtil

class FundAdapter(
    private val onFundClick: (Fund) -> Unit,
    private val onFundLongClick: (Fund) -> Unit = {}
) : ListAdapter<Fund, FundAdapter.FundViewHolder>(FundDiffCallback()) {

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): FundViewHolder {
        val binding = ItemFundBinding.inflate(
            LayoutInflater.from(parent.context), parent, false
        )
        return FundViewHolder(binding)
    }

    override fun onBindViewHolder(holder: FundViewHolder, position: Int) {
        holder.bind(getItem(position))
    }

    inner class FundViewHolder(
        private val binding: ItemFundBinding
    ) : RecyclerView.ViewHolder(binding.root) {

        init {
            binding.root.setOnClickListener {
                onFundClick(getItem(adapterPosition))
            }
            binding.root.setOnLongClickListener {
                onFundLongClick(getItem(adapterPosition))
                true
            }
        }

        fun bind(fund: Fund) {
            with(binding) {
                tvFundName.text = fund.name
                tvFundMeta.text = "${fund.totalMembers} members · ${CurrencyUtil.format(fund.monthlyAmount)}/mo"
                tvMonthly.text = CurrencyUtil.format(fund.monthlyAmount)
                tvInterestRate.text = "${fund.interestRate}% / mo"
                tvOverdueCount.text = fund.overdueCount.toString()

                applyDotColor(fund.dotColor)

                when {
                    fund.isClosed -> {
                        tvStatus.text = "Closed"
                        tvStatus.setBackgroundResource(R.drawable.pill_gray_bg)
                        tvStatus.setTextColor(root.context.getColor(R.color.closed_text))
                        tvClosedDate.visibility = View.VISIBLE
                        tvClosedDate.text = "Closed: ${fund.closedDate.orEmpty()}"
                        tvMonthly.text = "—"
                        tvInterestRate.text = "—"
                    }
                    fund.isOverdue -> {
                        tvStatus.text = "⚠ Overdue"
                        tvStatus.setBackgroundResource(R.drawable.pill_orange_bg)
                        tvStatus.setTextColor(root.context.getColor(R.color.color_orange))
                        tvClosedDate.visibility = View.GONE
                    }
                    else -> {
                        tvStatus.text = "✓ All Paid"
                        tvStatus.setBackgroundResource(R.drawable.pill_green_bg)
                        tvStatus.setTextColor(root.context.getColor(R.color.color_green))
                        tvClosedDate.visibility = View.GONE
                    }
                }
            }
        }

        private fun applyDotColor(colorHex: String) {
            runCatching {
                val c = Color.parseColor(colorHex)
                GradientDrawable().apply {
                    shape = GradientDrawable.OVAL
                    setColor(c)
                    binding.viewDot.background = this
                }
                GradientDrawable().apply {
                    shape = GradientDrawable.RECTANGLE
                    cornerRadius = 20f
                    setColor(Color.argb(34, Color.red(c), Color.green(c), Color.blue(c)))
                    setStroke(4, Color.argb(85, Color.red(c), Color.green(c), Color.blue(c)))
                    binding.frameDot.background = this
                }
            }
        }
    }

    private class FundDiffCallback : DiffUtil.ItemCallback<Fund>() {
        override fun areItemsTheSame(old: Fund, new: Fund) = old.id == new.id
        override fun areContentsTheSame(old: Fund, new: Fund) = old == new
    }
}
